importer/balanced/builder.go

```diff
@@ -1,113 +1,253 @@
+	// has reached its maximum capacity of `db.Maxlinks()` per node)
+	// extend it by making it a sub-DAG of a bigger DAG with `depth+1`.
+	for depth := 1; !db.Done(); depth++ {
+
+		// Add the old `root` as a child of the `newRoot`.
+		newRoot := db.NewFSNodeOverDag(ft.TFile)
+		newRoot.AddChild(root, fileSize, db)
+
+		// Fill the `newRoot` (that has the old `root` already as child)
+		// and make it the current `root` for the next iteration (when
+		// it will become "old").
+		root, fileSize, err = fillNodeRec(db, newRoot, depth)
+		if err != nil {
+			return nil, err
+		}
 	}
 
-	return out, nil
+	return db.AddNodeAndClose(root)
 }
 
-// fillNodeRec will fill the given node with data from the dagBuilders input
-// source down to an indirection depth as specified by 'depth'
-// it returns the total dataSize of the node, and a potential error
+// fillNodeRec will fill the given internal (non-leaf) `node` with data, either by
+// creating leaf data nodes (if depth is 1) or by creating internal nodes and
+// calling itself recursively on them until they are filled with data. The data to
+// fill the node with is provided by DagBuilderHelper.
+//
+// `node` represents a (sub-)DAG root that is being filled. If called recursively,
+// it is `nil`, a new node is created. If it has been called from `Layout` (see
+// diagram below) it points to the new root (that increases the depth of the DAG),
+// it already has a child (the old root). New children will be added to this new
+// root, and those children will in turn be filled (calling `fillNodeRec`
+// recursively).
+//
+//                      +-------------+
+//                      |   `node`    |
+//                      |  (new root) |
+//                      +-------------+
+//                            |
+//              +-------------+ - - - - - - + - - - - - - - - - - - +
+//              |                           |                       |
+//      +--------------+             + - - - - -  +           + - - - - -  +
+//      |  (old root)  |             |  new child |           |            |
+//      +--------------+             + - - - - -  +           + - - - - -  +
+//              |                          |                        |
+//       +------+------+             + - - + - - - +
+//       |             |             |             |
+//  +=========+   +=========+   + - - - - +    + - - - - +
+//  | Chunk 1 |   | Chunk 2 |   | Chunk 3 |    | Chunk 4 |
+//  +=========+   +=========+   + - - - - +    + - - - - +
+//
+// The `node` to be filled uses the `FSNodeOverDag` abstraction that allows adding
+// child nodes without packing/unpacking the UnixFS layer node (having an internal
+// `ft.FSNode` cache).
+//
+// It returns the `ipld.Node` representation of the passed `node` filled with
+// children and the `nodeFileSize` with the total size of the file chunk (leaf)
+// nodes stored under this node (to propagate it upwards in the DAG to its parent).
```

Comment:
Instead of:

> (to propagate it upwards in the DAG to its parent)

I might say:

> (parent nodes store this to enable efficient seeking through the DAG when reading data later)

I think it’s self-evident that the data is being propagated upward, and this describes a little more clearly why we want that information.